Docs extended: Installation behind firewall (#8034)
jupyterlab/jupyterlab: docs/source/getting_started/installation.rst

``jupyterlab/build`` directory manually if desired.

Installation behind firewall

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Installation problems
~~~~~~~~~~~~~~~~~~~~~
```

If your computer is behind corporate proxy or firewall,
you may encounter an HTTP and SSL errors due to custom security profiles managed by corporate IT departments.

> 

Author: Of course, see in updated PR

This is expected, because your company can block connection to widely used repositories in Python community.

> Good start is to ask your network administrator, to allow all http + https communication to widely used Python / Jupyter servers:

Reviewer: ```suggestion
Here are some widely-used sites that host packages in the Python and JavaScript open-source ecosystem. Your network adminstrator may be able to allow http and https connections to these:
```

- \*.pypi.org
- \*.pythonhosted.org
- \*.continuum.io
- \*.anaconda.com
- \*.conda.io
- \*.github.com
- \*.githubusercontent.com
- \*.npmjs.com
- \*.yarnpkg.com

Alternatively you can specify proxy user (mostly domain user with password),
that is allowed to communicate via network. This can be easily achieved

> by setting 2 common environment variables `HTTP_PROXY` and `HTTPS_PROXY`.

Reviewer: ```suggestion
by setting two common environment variables: `HTTP_PROXY` and `HTTPS_PROXY`.
```

> These variables are automatically used by many open-source tools (like ``conda``), if setup correctly.

Reviewer: ```suggestion
These variables are automatically used by many open-source tools (like ``conda``) if set correctly.
```

.. code:: bash

    # For Windows
    set HTTP_PROXY=http://USER:PWD@proxy.company.com:PORT
    set HTTPS_PROXY=https://USER:PWD@proxy.comp any.com:PORT

    # For Linux / MacOS
    export HTTP_PROXY=http://USER:PWD@proxy.company.com:PORT
    export HTTPS_PROXY=https://USER:PWD@proxy.company.com:PORT

In case you can communicate via HTTP, but installation with ``conda`` fails
on connectivity problems to HTTPS servers, you can disable using SSL for ``conda``.

.. warning:: Disable SSL in communication is generally not recommended and involves potential security risk.

>     But for testing purposes, this step can help you to get rid of all HTTPS problems in ``conda`` quickly.

Reviewer: ```suggestion
.. warning:: Disable SSL in communication is generally not recommended and involves potential security risk.
```

Reviewer: It seems this was not resolved? The purpose here was to not undercut the effectiveness of the warning.

Author: jj, clear now ;)
(for both occurences)

.. code:: bash

>     # Setup conda to not use SSL

Reviewer: ```suggestion
    # Set up conda to not use SSL
```

    conda config --set ssl_verify False

> You can do similar thing for ``pip``.

Reviewer: ```suggestion
You can do a similar thing for ``pip``.
```

The approach here is to mark repository servers as trusted hosts,
which means, SSL communication will not be required for downloading Python libraries.

.. code:: bash

    # Install pandas (without SSL)
    pip install --trusted-host pypi.org --trusted-host files.pythonhosted.org pandas

Using the tips from above, you can handle most installation problems

> related to Python libraries.

Reviewer: ```suggestion
Using the tips from above, you can handle many network problems 
related to installing Python libraries.
```

> Many Jupter extensions require to have working ``npm`` command,

Reviewer: ```suggestion
Many Jupter extensions require having a working ``npm`` command,
```

> which is required for downloading useful Jupyter extensions / other javascript dependencies.

Reviewer: ```suggestion
which is required for downloading useful Jupyter extensions or other JavaScript dependencies.
```

Reviewer: It seems this was not resolved?

.. code:: bash

    # Do not require SSL

>     npm set strict-ssl False

Reviewer: ```suggestion
```

Reviewer: Can you add this separately with a similar warning as above? We don't want to encourage people to disable ssl by default.

Author: Of course, see in updated PR

Reviewer: It seems that this was not resolved in the current PR?

Author: Warning added. Fixed now ;)

Reviewer: Thanks, that is better. Can you do one more thing, and move this disabling ssl section down below the proxy setting, like you have above for conda? My concern is that people reading straight through see the disabling ssl line first and think that is the first they should try, when really they should first try setting the proxy, and if that doesn't work, they may consider disabling ssl.

Reviewer: I think you did it well above, where you have the proxy config, then a sentence about how disabling ssl might help if there are still issues, then the warning, *then* the instructions (after the warning and explanatory note).

Author: Very good  idea - I see your point.
Please check - it should be done now.

    # Set proxy for NPM
    npm config set proxy http://USER:PWD@proxy.company.com:PORT
    npm config set proxy https://USER:PWD@proxy.company.com:PORT

>     # Set default registry for NPM (optional, useful in case if common javascript libs cannot be found)

Reviewer: ```suggestion
    # Set default registry for NPM (optional, useful in case if common JavaScript libs cannot be found)
```
